dll/win32/syssetup/wizard.c

```diff
@@ -2880,6 +2881,97 @@ ProcessUnattendSection(
     }
 }
 
+static VOID
+AddInstallationSource(
+    IN HKEY hKey,
+    IN LPWSTR lpPath)
+{
+    LONG res;
+    DWORD dwRegType;
+    DWORD dwPathLength = 0;
+    DWORD dwNewLength = 0;
+    LPWSTR Buffer = NULL;
+    LPWSTR Path;
+
+    res = RegQueryValueExW(
+        hKey,
+        L"Installation Sources",
+        NULL,
+        &dwRegType,
+        NULL,
+        &dwPathLength);
+
+    if (res != ERROR_SUCCESS ||
+        dwRegType != REG_MULTI_SZ ||
+        dwPathLength == 0 ||
+        dwPathLength % sizeof(WCHAR) != 0)
+    {
+        dwPathLength = 0;
+        goto set;
+    }
+
+    /* Reserve space for existing data + new string */
+    dwNewLength = dwPathLength + (wcslen(lpPath) + 1) * sizeof(WCHAR);
+    Buffer = HeapAlloc(GetProcessHeap(), 0, dwNewLength);
+    if (!Buffer)
+        return;
+
+    ZeroMemory(Buffer, dwNewLength);
+
+    res = RegQueryValueExW(
+        hKey,
+        L"Installation Sources",
+        NULL,
+        NULL,
+        (LPBYTE)Buffer,
+        &dwPathLength);
+
+    if (res != ERROR_SUCCESS)
+    {
+        HeapFree(GetProcessHeap(), 0, Buffer);
+        dwPathLength = 0;
+        goto set;
+    }
+
+    /* Sanity check, these should already be zeros */
+    Buffer[dwPathLength / sizeof(WCHAR) - 2] = UNICODE_NULL;
+    Buffer[dwPathLength / sizeof(WCHAR) - 1] = UNICODE_NULL;
+
+    for (Path = Buffer; *Path; Path += wcslen(Path) + 1)
+    {
+        /* Check if path is already added */
+        if (PathIsSameRootW(Path, lpPath) && wcslen(Path) == wcslen(lpPath))
+            goto cleanup;
```

Comment:
If anyone knows a better way of comparing paths, please send a suggestion. :smile:

Reply:
Normalize the paths, compare the components one by one.